Mask grades on progress page according to show_correctness setting

pomegranited opened the pull request:
This change ensures that *individual* problem scores are hidden from learners on the Progress page, when indicated by the "Results Visibility" setting on the subsection.  For subsections where results are hidden:
* detailed problem score breakdown is replaced by informational text
* subsection grades are ~~hidden~~ visible
* grades shown in the graph ~~do not~~ reflect hidden problem scores.

The Results Visibility setting was introduced by https://github.com/edx/edx-platform/pull/14737, and it has the following options:
* `always`: (default) scores and correctness are always shown to learners and staff members.
* `never`: scores and correctness are never shown to learners or staff members.  This may be useful for survey-type questions where correctness is not of interest.
* `past_due`: scores and correctness are shown to learners after the subsection due date has passed (or the course due date, for self-paced courses).  However, staff members can always see scores and correctness for `past_due` subsections.  This may be useful for exams or other assessment exercises.
    Also, if there is no due date set, then `past_due` subsections `always` show correctness.

The `past_due` setting is the tricky bit, but also critical to the feature.  Staff members must be able to run grade reports before the problems are due, so they can see the current grades for their students as the course is run.  Staff members can also masquerade as a student to view her Progress page, and see individual scores for `past_due` subsections. However learners are not be able to see their individual scores yet.

**JIRA tickets**: [OSPR-1762](https://openedx.atlassian.net/browse/OSPR-1762)

**Discussions**: [Technical discovery document](https://docs.google.com/document/d/1yGAbIl5lZUXB_BTFDp5dx75fEypQHA7Ipg7M9PATKaw/edit)

**Screenshots**: Updated versions will follow once [remaining inconsistency](https://github.com/edx/edx-platform/pull/15018#issuecomment-301009640) has been fixed.

<!--
Initial Progress page (no responses submitted):
![screen shot 2017-05-04 at 11 33 16 pm](https://cloud.githubusercontent.com/assets/7556571/25707283/31035864-3122-11e7-9706-6dad44ca005f.png)
![screen shot 2017-05-05 at 12 24 33 am](https://cloud.githubusercontent.com/assets/7556571/25709726/3ed2d9cc-3129-11e7-90c6-48675fa43e46.png)

Progress page after correct responses submitted for the first unit in each subsection:
![screen shot 2017-05-05 at 12 13 36 am](https://cloud.githubusercontent.com/assets/7556571/25709287/0f3c36dc-3128-11e7-9e32-1ac25bce5b60.png)
![screen shot 2017-05-05 at 12 23 43 am](https://cloud.githubusercontent.com/assets/7556571/25709688/25015352-3129-11e7-9e5c-9f20a91db231.png)
-->

**Sandbox URL**:

* LMS: https://pr15018.sandbox.opencraft.hosting/
* Studio: https://studio-pr15018.sandbox.opencraft.hosting/

**Partner information**: 3rd party-hosted open edX instance

**Deployment targets**: edx.org and edge.edx.org

**Merge deadline**: ASAP

**Testing instructions**:

Test courses:
* [Instructor-paced course](https://pr15018.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+CAPA1+2017/info) - [show_correctness.course.tar.gz](https://github.com/edx/edx-platform/files/976330/show_correctness.course.tar.gz)
* [Self-paced course](https://pr15018.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+CAPA2+2017/info) - [self_paced.show_correctness.course.tar.gz](https://github.com/edx/edx-platform/files/976331/self_paced.show_correctness.course.tar.gz) - need to [add self-paced configuration](https://pr15018.sandbox.opencraft.hosting/admin/self_paced/selfpacedconfiguration/) to allow self-paced courses.

Please see https://github.com/edx/edx-platform/pull/14737 for manual setup instructions.

The `honor` user on the  [Instructor-paced sandbox course](https://pr15018.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+CAPA1+2017/progress) has already performed the steps below.

1. Enroll in the test courses as a learner.
1. Visit the Progress page.  Note that:
     * Sections named "Show Correctness: Always" show a breakdown of practice scores.
     * Sections named "Show Correctness: Never" show "Practice scores are hidden."
     * Sections named "Show Correctness: Past Due" have a due date, and show "Practice scores are hidden until the due date." instead of the list of practice scores.
1. Visit each of the subsections, graded and ungraded, and complete some problems correctly.
1. Re-visit the Progress page, and note that:
     * Sections named "Show Correctness: Always" have updated scores, the Graded subsection scores are visible in the grades chart, and individual scores are visible (and up-to-date).
     * Sections named "Show Correctness: Never" or "Past Due" ~~have not changed~~ have updated scores, the Graded subsection scores are ~~*not*~~ visible in the grades chart, but problem-level scores are hidden.
     * The exception is the sections labeled "Legacy Problem setting".  Legacy problem settings show scores in the summary regardless of their "Results Visibility" setting -- this is because the legacy setting is at the problem level not the subsection, and so is not available in this view.
1. View the output of the Grades API, e.g. [`/api/grades/v0/course_grade/course-v1%3AOpenCraft%2BCAPA1%2B2017/users/`](https://pr15018.sandbox.opencraft.hosting/api/grades/v0/course_grade/course-v1%3AOpenCraft%2BCAPA1%2B2017/users/). Note that the `percent` grade reflected matches the Total grade visible on the Progress page.

1. Login to the LMS as `staff@example.com`.
1. View the Progress page.  Note that you can see the staff user's problem scores for all subsections except "Show Correctness: Never".
1. View course as a Specific Learner, and enter in your learner username.  Note that you can see the learner's individual problem scores for all subsections except "Show Correctness: Never".
1. View the output of the Grades API for that learner, e.g. [`/api/grades/v0/course_grade/course-v1%3AOpenCraft%2BCAPA1%2B2017/users/?username=honor`](https://pr15018.sandbox.opencraft.hosting/api/grades/v0/course_grade/course-v1%3AOpenCraft%2BCAPA1%2B2017/users/?username=honor).  Note that the `percent` grade reflected matches the Total grade visible on the masqueraded Progress page.

**Reviewers**
- [x] @itsjeyd 
- [x] @yro 
- [x] @nasthagiri

edx-webhook commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=15018&repo=edx%2Fedx-platform"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-1762](https://openedx.atlassian.net/browse/OSPR-1762) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

e-kolpakov commented:
@pomegranited This is indeed a too invasive change - the approach I had in my mind was to let grade factories do their thing and post-process grades in the view, just before rendering it.

Namely, `chapter_grades` [here](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/views/views.py#L868) is constructed [here](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/grades/new/course_grade.py#L67) - if I understand right it is a subsection_location => subsection_grade dictionary. So, the idea was to use the subsection_location to pull the subsection from the modulestore and check it's answer_available setting.

I'm not 100% sure if I traced it right though.

pomegranited commented:
@e-kolpakov Ah that'll be much easier..  I got caught in the rabbit hole.  Will fix this up and have it ready for a review on Monday. CC @itsjeyd

itsjeyd commented on lms/templates/courseware/progress.html:
```diff
@@ -154,22 +154,24 @@ <h3 class="hd hd-3" id="chapter_${loop.index}">${ chapter['display_name']}</h3>
+                                        if section.show_grades(staff_access):
+                                          earned = section.all_total.earned
+                                          total = section.all_total.possible
+                                          percentageString = ''
```
@pomegranited We should probably standardize on `snake_case` here.

pomegranited replied:
Oops, missed that one.. Fixed with https://github.com/edx/edx-platform/pull/15018/commits/cc37c0fbce1ac195717ac38723a3510021968308#diff-54c933c75934e410ed718cbb2345158aR162

pomegranited commented on common/lib/xmodule/xmodule/graders.py:
```diff
@@ -462,3 +463,38 @@ def _min_or_none(itr):
+    """
+
+    """
+    Constants use to indicate when to show correctness
```
Addressed by https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-743e6bcf9ca89e1c3a85c3ec2c9d81b5R477

pomegranited commented on common/lib/xmodule/xmodule/tests/test_graders.py:
```diff
@@ -315,3 +316,80 @@ def test_grader_with_invalid_conf(self, invalid_conf, expected_error_message):
+    @ddt.data(True, False)
+    def test_show_correctness_never(self, has_staff_access):
+        """
+        Test show_correctness="never" hides correctness from learners and course staff.
```
Addressed by https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-5b0879ec1e8d310b6a2d9f5c3ed9e779R361.

pomegranited commented on common/lib/xmodule/xmodule/tests/test_graders.py:
```diff
@@ -315,3 +316,80 @@ def test_grader_with_invalid_conf(self, invalid_conf, expected_error_message):
+        (ShowCorrectness.PAST_DUE, None, True, True),
+    )
+    @ddt.unpack
+    def test_show_correctness_past_due(self, show_correctness, due_date_str, has_staff_access, expected_result):
```
Good point.. copy-paste leftovers!

Addressed by https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-5b0879ec1e8d310b6a2d9f5c3ed9e779R368.

pomegranited commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -1653,6 +1669,255 @@ def mock_certificate_downloadable_status(
+    NOW = datetime.now(UTC)
+    DAY_DELTA = timedelta(days=1)
+    YESTERDAY = NOW - DAY_DELTA
+    TOMORROW = NOW + DAY_DELTA
```
Since `TODAY` is in the past, it behaves just like `YESTERDAY`.

Addressed with https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1682, https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1827, https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1827, etc.

pomegranited commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -1653,6 +1669,255 @@ def mock_certificate_downloadable_status(
+
+    def setup_course(self, show_correctness='', due_date=None, graded=False, add_problem=True, **course_options):
+        """
+        Set up course with a subsection with the given show_correctness, due_date, and graded settings.
```
Went one better, and separated it out into its own method:  https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1732

When I started writing these tests, I wasn't very familiar with how the modulestore test objects worked, and so didn't know how to update an existing course.  But lo, there's a way!

pomegranited commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -1653,6 +1669,255 @@ def mock_certificate_downloadable_status(
+            self.problem.scope_ids.usage_id,
+            field_data_cache,
+        )._xmodule
+        module.system.publish(self.problem, 'grade', grade_dict)
```
Addressed with https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1765.

pomegranited commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -1653,6 +1669,255 @@ def mock_certificate_downloadable_status(
+                                    label="Homework-grade_breakdown"))
+
+        ungraded_data = [dict(color='#b72121', label='Homework',
+                              data=[[1, 0.0], [2, 0.0], [3.25, 0.0]])]
```
Did the same for `"Homework"`, addressed with https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1777.

pomegranited commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -1653,6 +1669,255 @@ def mock_certificate_downloadable_status(
+        (ShowCorrectness.PAST_DUE, None, False),
+        (ShowCorrectness.PAST_DUE, None, True),
+        (ShowCorrectness.NEVER, None, False),
+        (ShowCorrectness.NEVER, None, True),
```
Sure, might as well!  Addressed with https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1835.

pomegranited commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -1653,6 +1669,255 @@ def mock_certificate_downloadable_status(
+    @ddt.unpack
+    def test_progress_page_no_problem_scores(self, show_correctness, due_date, graded):
+        """
+        Test that "no problem scores are present" for a course with no problems in, regardless of the various show correctness settings
```
Addressed with https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1852.

pomegranited commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -1653,6 +1669,255 @@ def mock_certificate_downloadable_status(
+    @ddt.unpack
+    def test_progress_page_hide_scores_from_learner(self, show_correctness, due_date, graded, show_grades):
+        """
+        Test that problem scores are hidden on progress page when correctness is not available to the learner.
```
Addressed with https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1891.

pomegranited commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -1653,6 +1669,255 @@ def mock_certificate_downloadable_status(
+        Test that problem scores are hidden on progress page when correctness is not available to the learner.
+        """
+        self.setup_course(show_correctness=show_correctness, due_date=due_date, graded=graded, add_problem=True)
+        resp = self._get_progress_page()
```
Technically no, as it's covered by the [base class `setUp()`](https://github.com/open-craft/edx-platform/blob/29b9ad4e03294b8876eec28ed85b5e52e5891ddb/lms/djangoapps/courseware/tests/test_views.py#L1188), but I've added it in anyway for symmetry: https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1897

pomegranited commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -1653,6 +1669,255 @@ def mock_certificate_downloadable_status(
+    @ddt.unpack
+    def test_progress_page_hide_scores_from_staff(self, show_correctness, due_date, graded, show_grades):
+        """
+        Test that problem scores are hidden from staff viewing a learner's progress page only if show_correctness=never
```
Addressed with https://github.com/edx/edx-platform/pull/15018/commits/d7e44c50acc6dc3a73fc400f410cfe4ceaca834d#diff-8aefe502dfa8ee5a3e18c9227e61b034R1942.

pomegranited commented:
@itsjeyd I've addressed your nits, and when the tests pass, I'll spawn a new sandbox.

I've also taken @e-kolpakov 's suggestion on board with 911a197, and changed the `CourseGrade.view_as_staff = True` by default, to `CourseGrade.view_as_learner = False` by default.  I agree it's cleaner, and hopefully less confusing.

pomegranited commented:
jenkins run js

pomegranited commented:
jenkins run bokchoy

itsjeyd commented:
@pomegranited Thanks for the updates! I made another pass over the code, and it looks good to me. Will formally approve the changes once I've tested them on the sandbox.

itsjeyd commented:
@pomegranited I tested the changes on the sandbox (using `audit` as a non-staff learner). Got expected results for all steps except one:

> Re-visit the Progress page, and note that: ...
> Sections named "Show Correctness: Never" or "Past Due" have not changed, **and the Graded subsection scores are not visible in the grades chart**.

That last part doesn't seem to be working. When logged in as `audit` I can see my results for HW3 in the grades chart:

![hw3-visible](https://cloud.githubusercontent.com/assets/961441/25845039/ac41bc68-34ac-11e7-815c-b02019c61bb3.png)

---

Another thing I wanted to check was whether the Progress page would show results for "Show Correctness: Past Due" after setting the subsection due date to a date in the past. It didn't work; the LMS didn't pick up the due date. That behavior is probably unrelated to your changes. What steps did you take to set the original due date? I just clicked the cog on the subsection in Studio, changed the date, and hit "Save". Studio didn't offer an option to publish the changes afterwards.

e-kolpakov commented:
@itsjeyd I believe this might have something to do with [this comment](https://docs.google.com/document/d/1P9kerbwZRW7sFBC46pfncQs1nlu8tx6KsbyF1mEDX74/edit?disco=AAAAA9UAULA) - basically MIT PM and @sstack22 decided to only hide problem scores, but not subsection scores or grades in the graph.

So, there's probably other bug - scores should be shown for "Never" and "Past Due" sections, but not for individual problems in them.

itsjeyd commented:
@cahrens @nasthagiri This is ready for you to have a look. We have one inconsistency left to sort out, which is that the Progress page currently doesn't show scores for subsections that have "Show Correctness" set to "Never" or "Past Due", cf. https://github.com/edx/edx-platform/pull/15018#issuecomment-300639485. We don't have the bandwidth to fix that this week, but we don't anticipate the fix to trigger fundamental changes to the general approach, and we'd like to get your input on the existing implementation. Thanks!

e-kolpakov commented:
@itsjeyd You might want to approve the PR (or at least revoke your "request changes" review)

cahrens commented:
@e-kolpakov and @itsjeyd This ticket is still marked as "Waiting on Author". Is that the correct status? I'm also not clear what comment this refers to, as the link isn't working:

"We have one inconsistency left to sort out, which is that the Progress page currently doesn't show scores for subsections that have "Show Correctness" set to "Never" or "Past Due", cf. #15018 (comment)."

itsjeyd commented:
@cahrens 

> This ticket is still marked as "Waiting on Author". Is that the correct status? 

Yes and no: We need to fix that last inconsistency (I'm planning on having a look at it soon), but we'd still like you to start looking at the existing changes, to see if the general approach looks good to you.

> I'm also not clear what comment this refers to, as the link isn't working: 

Sorry about that. The link seems to be working for me: https://github.com/edx/edx-platform/pull/15018#issuecomment-300639485. It should take you to the second-to-last comment that @e-kolpakov posted. Quoting him here for reference:

> @itsjeyd I believe this might have something to do with this comment - basically MIT PM and @sstack22 decided to only hide problem scores, but not subsection scores or grades in the graph.
> So, there's probably other bug - scores should be shown for "Never" and "Past Due" sections, but not for individual problems in them.

srpearce commented:
@pomegranited It looks like the main change that the docs will need when this PR merges is this:

[Set Problem Results Visibility](http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/developing_course/course_subsections.html#set-problem-results-visibility)
**Current text**
Show results when subsection is past due: For learners, results are not visible until the subsection due date (for instructor-paced courses) or the course end date (for self-paced courses) has passed. For course staff, results are always visible unless the staff member is previewing or viewing the course as a learner.

**New text**
Show results when subsection is past due: For learners, results are not visible until the subsection due date (for instructor-paced courses) or the course end date (for self-paced courses) has passed. For course staff, results are always visible, even when the staff member is previewing or viewing the course as a learner. 

Is that correct?

itsjeyd commented:
@srpearce If the [Set Problem Results Visibility](http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/developing_course/course_subsections.html#set-problem-results-visibility) section is about the "Progress" page (I wasn't sure because it doesn't mention the "Progress" page at all), the documentation updates might need to be more fine-grained. This is because results visibility settings now affect subsection scores and problem scores differently:

* For staff (and staff viewing "Progress" page as specific learner):
    * Sections with "Show Correctness: Always"
        * show a breakdown of practice scores.
        * show subsection scores, and subsection scores are visible in the grades chart (for graded subsections).
    * Sections with "Show Correctness: Never"
        * show "Practice scores are hidden." (or "Problem scores are hidden." for graded subsections).
        * show subsection scores, and subsection scores are visible in the grades chart (for graded subsections).
    * Sections with "Show Correctness: Past Due"
        * have a due date
        * show a breakdown of practice scores.
        * show subsection scores, and subsection scores are visible in the grades chart (for graded subsections).
* For learners:
    * Sections with "Show Correctness: Always": **same behavior as for staff users**
    * Sections with "Show Correctness: Never":  **same behavior as for staff users**
    * Sections with "Show Correctness: Past Due"
        * have a due date
        * show "Practice scores are hidden until the due date." (or "Problem scores are hidden until the due date." for graded subsections).
        * show subsection scores, and subsection scores are visible in the grades chart (for graded subsections).

pomegranited commented:
@itsjeyd Thanks for clarifying the feature with edX, and fixing up the PR description to suit the desired outcome.

I've reverted back to 62cb9de, removed the code that alters the subsection/course grades shown, and updated the tests to only check for the hidden individual scores on the progress page.

Have also re-applied changes to address the nits from your previous PR reviews.

Ready for you to re-review!

pomegranited commented:
jenkins run bokchoy

cahrens commented:
Folks with more experience in grading are going to review this PR. The OSPR ticket is in Team Neem's current sprint.

pomegranited commented:
Thanks for your help and coordination, @cahrens !

yro commented on lms/templates/courseware/progress.html:
```diff
@@ -180,12 +180,30 @@ <h4 class="hd hd-4">
-                                            %endfor
-                                        </dl>
+                                          %if section.show_grades(staff_access):
+                                          <dl class="scores">
```
Is there a use case for granular staff/non-staff access users? Why not use the template global 'staff_access'? 

`section.show_grades(staff_access)` seems to be defaulting to True for me on devstack -- I believe the variable exists regardless of truthiness, but it's possible to be less granular per module in this instance (I think -- push back if I'm wrong, and I often am).

itsjeyd replied:
@yro (Taking over for @pomegranited since she's away.)

IIUC, you're asking if it would be possible to do

```
%if staff_access:
```

instead of

```
%if section.show_grades(staff_access):
```

here.

Doing that wouldn't give us the behavior that we want: If a subsection is set up to never show correctness, we want to [hide individual scores from all users](https://github.com/edx/edx-platform/pull/15018/files#diff-743e6bcf9ca89e1c3a85c3ec2c9d81b5R488), including those that have staff access, cf. https://github.com/edx/edx-platform/pull/15018#issuecomment-302102287.

This behavior mimics what we're doing for the "Show answer" setting (see the [definition](https://github.com/edx/edx-platform/pull/15018/files#diff-0c9995bbce7571d2d55664680e485452L892) of `CapaMixin.answer_available`), and was approved in #14737 ([comment](https://github.com/edx/edx-platform/pull/14737/files#r110754178)), the PR that first introduced the "Show correctness" setting.

yro replied:
@itsjeyd Right on, thanks for the clarification.

yro commented on lms/djangoapps/grades/transformer.py:
```diff
@@ -37,7 +38,10 @@ class GradesTransformer(BlockStructureTransformer):
+    FIELDS_TO_COLLECT = [
+        u'due', u'format', u'graded', u'has_score', u'weight', u'course_version', u'subtree_edited_on',
+        u'show_correctness',
+    ]
```
Nit: make newlines for elements of array

yro commented on common/lib/xmodule/xmodule/tests/test_graders.py:
```diff
@@ -315,3 +316,80 @@ def test_grader_with_invalid_conf(self, invalid_conf, expected_error_message):
+        """
+        self.assertFalse(ShowCorrectness.correctness_available(show_correctness=ShowCorrectness.NEVER,
+                                                               has_staff_access=has_staff_access))
+
```
Nit: newlines

yro commented on common/lib/xmodule/xmodule/tests/test_graders.py:
```diff
@@ -315,3 +316,80 @@ def test_grader_with_invalid_conf(self, invalid_conf, expected_error_message):
+        """
+        self.assertTrue(ShowCorrectness.correctness_available(show_correctness=show_correctness,
+                                                              has_staff_access=has_staff_access))
+
```
Nit: newlines

yro commented on common/lib/xmodule/xmodule/tests/test_graders.py:
```diff
@@ -315,3 +316,80 @@ def test_grader_with_invalid_conf(self, invalid_conf, expected_error_message):
+            due_date = None
+        else:
+            due_date = getattr(self, due_date_str)
+        self.assertEquals(ShowCorrectness.correctness_available(ShowCorrectness.PAST_DUE, due_date, has_staff_access),
```
Nit: newlines

yro commented:
@pomegranited This is really nice! I just had some nits and a question about the progress page, but functionally this is great stuff. Thanks!

itsjeyd commented:
jenkins run bokchoy

itsjeyd commented:
Rebased from b6adc8e29e9278089cd1851c86d860822f423eef.

itsjeyd commented:
Squashed from ff9a3d920703b7569dc74fb414bc64b5c9f57322.

itsjeyd commented:
jenkins run a11y

itsjeyd commented:
jenkins run quality

itsjeyd commented:
jenkins run js

itsjeyd commented:
jenkins run python

itsjeyd commented:
@yro @nasthagiri Thanks for reviewing! I rebased and squashed, and will merge once the build is green.

itsjeyd commented:
jenkins run bokchoy

e-kolpakov commented:
jenkins run bokchoy

itsjeyd commented:
jenkins run bokchoy

e-kolpakov commented:
@itsjeyd @pomegranited Green build! Who's going to take privilege of merging it (certainly not me - there aren't a single line of code from me :))?

itsjeyd commented:
@e-kolpakov \o/

Since @pomegranited is on vacation I'm not sure if/when she'll see the recent updates. So I'm going to go ahead and flip the switch now :)
